File: crates/core/tests/fixtures/extractors/python/accounts.py
"""User accounts service.

This module owns the business logic for the user resource. Persistence is
abstracted behind the :class:`UserRepository` Protocol so the service can
be exercised against the in-memory implementation in tests and against a
real database in production.

The service does NOT validate input formats — that's the controller's
job, run via Pydantic before we even get here.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass, field
from typing import Iterable, Iterator, Optional, Protocol

# Module-level logger. Configured by the application; tests capture via caplog.
log = logging.getLogger(__name__)


@dataclass(frozen=True)
class User:
    """Immutable user record.

    The `tags` list is stored as an ordinary list (not a tuple) so callers
    that build users by hand don't have to convert; the `frozen=True` only
    prevents reassignment of the fields, not mutation of `tags`. Treat
    `tags` as conceptually frozen anyway.
    """

    id: str
    email: str
    name: str
    tags: list[str] = field(default_factory=list)


class UserNotFoundError(Exception):
    """Raised when a lookup by id returns nothing.

    Surfaced by the HTTP layer as 404. The id is exposed on the exception
    so handlers can include it in the response body.
    """

    def __init__(self, user_id: str) -> None:
        super().__init__(f"user {user_id} not found")
        self.user_id = user_id


class ConflictError(Exception):
    """Raised when a unique constraint would be violated."""

    def __init__(self, field_name: str, value: str) -> None:
        super().__init__(f"conflict on {field_name}={value!r}")
        self.field_name = field_name
        self.value = value


class UserRepository(Protocol):
    """Storage interface — see ``InMemoryRepository`` for the canonical impl."""

    def find_by_id(self, user_id: str) -> Optional[User]: ...
    def find_by_email(self, email: str) -> Optional[User]: ...
    def insert(self, user: User) -> User: ...
    def scan(self) -> Iterator[User]: ...


class UserService:
    """Business logic, injected with a repository at construction time."""

    def __init__(self, repo: UserRepository) -> None:
        # Underscore prefix marks this as internal; the rest of the codebase
        # accesses the repo only via service methods.
        self._repo = repo

    def get(self, user_id: str) -> User:
        user = self._repo.find_by_id(user_id)
        if user is None:
            raise UserNotFoundError(user_id)
        return user

    def create(self, email: str, name: str, tags: Iterable[str] = ()) -> User:
        # Uniqueness check before insert. The DB also has a unique index but
        # we want the structured ConflictError, not whatever the driver raises.
        if self._repo.find_by_email(email) is not None:
            raise ConflictError("email", email)
        new_user = User(id=make_id(email), email=email, name=name, tags=list(tags))
        log.info("creating user", extra={"email": email})
        return self._repo.insert(new_user)

    def with_tag(self, tag: str) -> list[User]:
        # Materialised list is fine — the tag-search endpoint paginates upstream.
        return [u for u in self._repo.scan() if tag in u.tags]


class InMemoryRepository:
    """Reference implementation. Thread-unsafe — wrap in a lock for prod use."""

    def __init__(self) -> None:
        self._by_id: dict[str, User] = {}

    def find_by_id(self, user_id: str) -> Optional[User]:
        return self._by_id.get(user_id)

    def find_by_email(self, email: str) -> Optional[User]:
        for user in self._by_id.values():
            if user.email == email:
                return user
        return None

    def insert(self, user: User) -> User:
        # TODO: race condition between this check and the assignment below;
        # acceptable for the in-memory test impl but real impls must use
        # `INSERT ... ON CONFLICT` or a transaction.
        if self.find_by_email(user.email) is not None:
            raise ConflictError("email", user.email)
        self._by_id[user.id] = user
        return user

    def scan(self) -> Iterator[User]:
        # Yielding lazily means callers can short-circuit without copying.
        yield from self._by_id.values()


def make_id(email: str) -> str:
    """Derive a stable id from the local part of the email.

    Falls back to the full address when no ``@`` is present; the schema
    upstream rejects that case but defending here keeps the function total.
    """
    head, _, _ = email.partition("@")
    return head.lower() if head else email.lower()

File: crates/core/tests/fixtures/extractors/go/accounts.go
// Package store contains the user-account persistence layer.
//
// The repository abstraction lets the HTTP handlers stay decoupled from
// the concrete storage. The in-memory implementation here is used by tests
// and by the local dev server; production wires the same interface to
// pgxpool in cmd/server/main.go.
package store

import (
	"errors"
	"fmt"
	"strings"
	"sync"
)

// User is the canonical record. Fields are tag-less because the JSON
// envelope is built by the HTTP layer, not here.
type User struct {
	ID    string
	Email string
	Name  string
	Tags  []string
}

// ErrNotFound is returned by lookup methods when no row matches.
// Callers should compare with errors.Is, not ==.
var ErrNotFound = errors.New("not found")

// ConflictError signals a uniqueness violation. The fields are exported so
// HTTP middleware can render structured 409 bodies.
type ConflictError struct {
	Field string
	Value string
}

// Error implements the error interface.
func (e *ConflictError) Error() string {
	return fmt.Sprintf("conflict on %s=%q", e.Field, e.Value)
}

// UserRepository is the contract every storage backend must satisfy.
//
// Implementations are expected to be safe for concurrent use; the HTTP
// runtime calls from many goroutines.
type UserRepository interface {
	FindByID(id string) (*User, error)
	FindByEmail(email string) (*User, error)
	Insert(user User) (*User, error)
	// Scan iterates over every user. Order is unspecified.
	Scan() <-chan User
}

// InMemory is a sync.RWMutex-backed reference implementation. Suitable
// for tests and local dev only — there is no persistence across restarts.
type InMemory struct {
	mu    sync.RWMutex
	users map[string]User
}

// NewInMemory builds an empty store.
func NewInMemory() *InMemory {
	return &InMemory{users: make(map[string]User)}
}

func (m *InMemory) FindByID(id string) (*User, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()
	u, ok := m.users[id]
	if !ok {
		return nil, ErrNotFound
	}
	// Return a copy so callers can't mutate our internal state.
	return &u, nil
}

func (m *InMemory) FindByEmail(email string) (*User, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()
	for _, u := range m.users {
		if u.Email == email {
			return &u, nil
		}
	}
	return nil, ErrNotFound
}

func (m *InMemory) Insert(user User) (*User, error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	// Linear uniqueness check is fine for the in-memory impl. Postgres
	// impl relies on the UNIQUE index on (email).
	for _, existing := range m.users {
		if existing.Email == user.Email {
			return nil, &ConflictError{Field: "email", Value: user.Email}
		}
	}
	m.users[user.ID] = user
	return &user, nil
}

// Scan ships every user over a channel that closes when iteration finishes.
// The producer goroutine releases the read lock as soon as it returns; the
// consumer drives back-pressure naturally by not reading.
func (m *InMemory) Scan() <-chan User {
	out := make(chan User)
	go func() {
		defer close(out)
		m.mu.RLock()
		defer m.mu.RUnlock()
		for _, u := range m.users {
			out <- u
		}
	}()
	return out
}

// WithTag returns every user wearing tag.
//
// TODO(perf): switch to a tag-index when the user count grows past O(10k);
// linear scan is currently fine for our small tenants.
func WithTag(repo UserRepository, tag string) []User {
	var result []User
	for u := range repo.Scan() {
		for _, t := range u.Tags {
			if t == tag {
				result = append(result, u)
				break
			}
		}
	}
	return result
}

// MakeID returns the local part of the email, lowercased. Falls back to
// the full address when no '@' is present (defensive).
func MakeID(email string) string {
	at := strings.IndexByte(email, '@')
	if at <= 0 {
		return strings.ToLower(email)
	}
	return strings.ToLower(email[:at])
}
